### PR TITLE
Added sequence iterator

### PR DESCRIPTION
**Pull request by luisenp**
## Types of changes
Added an iterator class that returns sequences of transitions.

This also deprecates method `ReplayBuffer.get_iterators()` and replaces it with a utility function in `mbrl.util.common` package. 

- [x] Docs change / refactoring / dependency upgrade
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue
This is a core component of #88  and #86. 

## How Has This Been Tested (if it applies)
I added unit tests for the iterator and the new utilities added, which check that the trajectories returned are consistent with the trajectories stored in the replay buffer.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](../../CONTRIBUTING.md) document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: tests/core/test_common_utils.py [hunk 1]
```diff
@@ -240,3 +240,59 @@ def test_populate_replay_buffer_collect_trajectories():
...
+            buffer.add(dummy, dummy, dummy, k, False)
+            k += 1
+        buffer.close_trajectory()
+
+    train_iter, val_iter = mbrl.util.common.get_sequence_buffer_iterator(
+        buffer, 32, 0.1, 10, 3
+    )
+    # For trajectories of length 20 and sequence length 10, there are
+    # 10 possible start states.
+    # There are 30 trajectories in total, so 10% is 3 trajectories
```
Comment (freiberg-roman) [inline tests/core/test_common_utils.py:286]: I really like this version, much better than what I did. I am already starting to make the changes in my PR. I think there might be a logical error. If you assume a trajectory from 0-19 with length 20, then there would be 11 valid starts from 0 to 10 (inclusive) and therefore 10 + 1 valid sequences.

This test also fails if the sequences are of length one.
```python
def test_get_sequence_buffer_iterators_with_length_one():  # edge case
    buffer = mbrl.util.replay_buffer.ReplayBuffer(
        1000, (1,), (1,), max_trajectory_length=1
    )
    dummy = np.ones(1)
    num_trajectories_train = 27
    num_trajectories_val = 3
    k = 0
    for i in range(num_trajectories_train):
        buffer.add(dummy, dummy, dummy, k, False)
        buffer.close_trajectory()

    for i in range(num_trajectories_val):
        buffer.add(dummy, dummy, dummy, k, False)
        buffer.close_trajectory()

    train_iter, val_iter = mbrl.util.common.get_sequence_buffer_iterator(
        buffer, 32, 0.1, 1, 3
    )

    assert len(train_iter._valid_starts) == 27  # fails in current implementation since length is 0
    assert len(val_iter._valid_starts) == 3  # fails in current implementation
```
This is simply fixed with
```python
valid_starts.extend(list(range(start, end - sequence_length + 1)))  # line 260 in replay_buffer.py
```
I think the other tests may need to be revised

**Review by freiberg-roman — CHANGES_REQUESTED** [review]

File: tests/core/test_common_utils.py (hunk 1 above)
Comment (luisenp) [reply tests/core/test_common_utils.py:286]: Oh right, current version is missing that `range()`'s end index is not inclusive. The funny thing is that I wrote the other test and counted 11 possible start states (instead of 10) but when my test failed I convinced myself that 10 was correct :) Good catch!

File: mbrl/util/common.py [hunk 2]
```diff
@@ -190,6 +194,116 @@ def create_replay_buffer(
...
+        val_iterator = SequenceTransitionIterator(
+            transitions,
+            val_trajectories,
+            batch_size,
+            sequence_length,
+            1,
+            rng=replay_buffer._rng,
+            max_batches_per_loop=max_batches_per_loop,
+        )
+        val_iterator.toggle_bootstrap()
```
[the remark this answers is not in the document]
Comment (luisenp) [reply mbrl/util/common.py:302]: True, it's not really necessary since in the latest version the iterator always sets `it._bootstrap_iter = False` if the ensemble size is 1. But, unless I'm missing something , this toggle shouldn't change anything, I'm curious why you had to comment it out.

[the remark this answers is not in the document]
Comment (freiberg-roman) [reply mbrl/util/common.py:302]: I tried to reuse the evaluate function for training and needed to disable both. The double toggle made it weird.
I think this should be already in the latest commit in #88 
```python
   def evaluate(self, dataset: TransitionIterator) -> torch.Tensor:
        """Evaluates the model on the validation dataset.

        Iterates over the dataset, one batch at a time, and calls
        :meth:`mbrl.models.Model.eval_score` to compute the model score
        over the batch. The method returns the average score over the whole dataset.

        Args:
            dataset (bool): the transition iterator to use.

        Returns:
            (tensor): The average score of the model over the dataset (and for ensembles, per
                ensemble member).
        """
        if isinstance(dataset, BootstrapIterator):
            dataset.toggle_bootstrap()
        if isinstance(dataset, SequenceTransitionIterator):
            dataset.toggle_sequenced()

        batch_scores_list = []
        for batch in dataset:
            avg_batch_score = self.model.eval_score(batch)
            batch_scores_list.append(avg_batch_score)
        batch_scores = torch.cat(batch_scores_list, axis=batch_scores_list[0].ndim - 2)

        if isinstance(dataset, BootstrapIterator):
            dataset.toggle_bootstrap()
        if isinstance(dataset, SequenceTransitionIterator):
            dataset.toggle_sequenced()

        mean_axis = 1 if batch_scores.ndim == 2 else (1, 2)
        batch_scores = batch_scores.mean(axis=mean_axis)

        return batch_scores`

[the remark this answers is not in the document]
Comment (luisenp) [reply mbrl/util/common.py:302]: Do you get an error if you don't remove the toggle from the common utility? If so, can you paste the error message here? I'm trying to figure out if there is a bug, since I would have expected toggle to do nothing for the validation iterator with ensemble size 1.

[the remark this answers is not in the document]
Comment (freiberg-roman) [reply mbrl/util/common.py:302]: If you go to #88, use
`algorithm=pddm dynamics_model=gaussian_mmlp_ensemble overrides=pddm_cartpole overrides.validation_ratio=0.2` settings and comment in the toggle you should see something similar to
`...
 File "/home/fiber/projects/mbrl-lib/mbrl/models/gaussian_mlp.py", line 340, in eval_score
    assert model_in.ndim == 2 and target.ndim == 2
AssertionError`

[the remark this answers is not in the document]
Comment (luisenp) [reply mbrl/util/common.py:302]: Oh, it looks like your version of `BootstrapIterator` is not up to date, check out [this](https://github.com/facebookresearch/mbrl-lib/blob/sequence_iterator/mbrl/util/replay_buffer.py#L170) line, which is missing in #88. While we finalize this PR, can you make the target of #88 be the `sequence_iterator` branch to make it easier to spot any discrepancies?

[the remark this answers is not in the document]
Comment (freiberg-roman) [reply mbrl/util/common.py:302]: Ok my fault, got it running locally

**Review by freiberg-roman — APPROVED** [review]

File: mbrl/util/replay_buffer.py [hunk 3]
```diff
@@ -151,24 +164,128 @@ def __next__(self):
...
+                "than the specified length."
+            )
+        # no need to pass transitions to super(), since it's only used by __getitem__,
+        # which this class replaces. Passing the set of possible starts allow us to
+        # use all the indexing machinery of the superclasses.
+        super().__init__(
+            self._valid_starts,  # type: ignore
+            batch_size,
+            ensemble_size,
+            shuffle_each_epoch=False,
```
Comment (JanS97) [inline mbrl/util/replay_buffer.py:243]: Is there a reason why ``shuffle_each_epoch=False`` is hardcoded here? I would prefer being able to pass shuffle_each_epoch as an argument to the constructor of SequenceTransitionIterator.
I tried this out by setting ``_shuffle_each_epoch = True`` from outside and it seemed to work fine.

File: mbrl/util/common.py [hunk 4]
```diff
@@ -190,6 +194,116 @@ def create_replay_buffer(
...
+        batch_size,
+        sequence_length,
+        ensemble_size,
+        rng=replay_buffer._rng,
+        max_batches_per_loop=max_batches_per_loop,
+    )
+
+    val_iterator = None
+    if val_size > 0:
+        val_trajectories = replay_buffer.trajectory_indices[train_size:]
```
Comment (JanS97) [inline mbrl/util/common.py:292]: I think it is better to sample the trajectories for the training/validation sets. In the current implementation, the first trajectories in the buffer are used as training set and the last trajectories as validation set. If the policy changed between the rollouts that generated these trajectories, there will be a distribution mismatch between the training and validation set. 
Furthermore, this will cause that the training algorithm does not see new data until it is somewhat old (until the validation set is filled with newer data). I expect that this slows down the training of many algorithms.

Sampling the trajectories should reduce the mismatch and allow the training algorithm to get new data more quickly.

```suggestion
    trajectory_indices = random.sample(replay_buffer.trajectory_indices, len(replay_buffer.trajectory_indices))
    train_trajectories = trajectory_indices[:train_size]

    train_iterator = SequenceTransitionIterator(
        transitions,
        train_trajectories,
        batch_size,
        sequence_length,
        ensemble_size,
        rng=replay_buffer._rng,
        max_batches_per_loop=max_batches_per_loop,
    )

    val_iterator = None
    if val_size > 0:
        val_trajectories = trajectory_indices[train_size:]
```

File: mbrl/util/common.py [hunk 5]
```diff
@@ -190,6 +194,116 @@ def create_replay_buffer(
...
+        val_iterator = SequenceTransitionIterator(
+            transitions,
+            val_trajectories,
+            batch_size,
+            sequence_length,
+            1,
+            rng=replay_buffer._rng,
+            max_batches_per_loop=max_batches_per_loop,
```
Comment (JanS97) [inline mbrl/util/common.py:300]:
```suggestion
            max_batches_per_loop=math.ceil(val_ratio * max_batches_per_loop),
```
In the current implementation ``val_ratio`` does not influence how many samples are used for evaluation if ``max_batches_per_loop`` is used (and the replay buffer contains enough data). Evaluation is then always done with ``max_batches_per_loop`` samples. With the changes above, one has more control over how many samples are actually used for evaluation. The value is rounded up to ensure that always at least one sample is used.

File: mbrl/util/replay_buffer.py (hunk 3 above)
Comment (luisenp) [reply mbrl/util/replay_buffer.py:243]: Ah, good catch! At some point while I was working on this I had a version where this would have messed up the trajectories, but this is no longer relevant as it only shuffles the order of start states. I'll add back this option.

File: mbrl/util/common.py (hunk 4 above)
Comment (luisenp) [reply mbrl/util/common.py:292]: I agree, this was not intentional behavior, but an actual bug. There is a permutation missing that was in the original [code](https://github.com/facebookresearch/mbrl-lib/blob/master/mbrl/util/replay_buffer.py#L419). This bug also happens in the new `get_basic_buffer_iterators`. Thanks for catching!

File: mbrl/util/common.py (hunk 5 above)
Comment (luisenp) [reply mbrl/util/common.py:300]: I'll think a bit more about this (whether to incorporate this change or add a separate option), but I agree there should be more flexibility here.

Comment (JanS97) [reply mbrl/util/common.py:300]: A separate option would also be fine for me

File: mbrl/util/replay_buffer.py (hunk 3 above)
Comment (JanS97) [reply mbrl/util/replay_buffer.py:243]: Perfect, thanks

Comment (luisenp) [reply mbrl/util/replay_buffer.py]: Added in https://github.com/facebookresearch/mbrl-lib/pull/91/commits/029e0ab84bb9bf9f41da980305dde085b3bb74ff

File: mbrl/util/common.py (hunk 4 above)
Comment (luisenp) [reply mbrl/util/common.py]: Fixed in https://github.com/facebookresearch/mbrl-lib/pull/91/commits/c02b702c2ce5596e2a18d950c706145e846d41f9. I realized there was no bug in `get_basic_buffer_iterators` because I  had added an option `replay_buffer.get_all(shuffle=True)`.

File: mbrl/util/common.py (hunk 5 above)
Comment (luisenp) [reply mbrl/util/common.py]: Added a separate option in https://github.com/facebookresearch/mbrl-lib/pull/91/commits/892534118d773c91461ec1fcc059d9f60ce532b4.
